File: src/fprotocol/fprotocol.py
import logging
logging.basicConfig(
    level=logging.WARN,
    format='%(levelname)s:%(filename)s:%(funcName)s:%(lineno)d->%(message)s',
)
logger = logging.getLogger(__name__)


import struct

class DynamicStruct:
    def __init__(self, field_defs):
        """
        field_defs: list of tuples (field_name, format_char)
        Example: [("id", "H"), ("payload", "1024s")]
        """
        self.fields = field_defs
        self.format = ''.join(fmt for _, fmt,_ in field_defs)
        self.size = struct.calcsize(self.format)
        self.values = {name: default_value for name, _,default_value in field_defs}
        self.callback = None


    def parse(self, data: bytes):
        index = 0
        next_array_size = 0
        total_len = len(data)
        
        for field in self.fields:
            field_name, fmt, _ = field
            field_size = struct.calcsize(fmt)

            
            # 原有解析逻辑
            if field_name.startswith("_"):
                self.values[field_name] = struct.unpack(fmt, data[index:index+field_size])[0]
                next_array_size = self.values[field_name]
                index += field_size
            else:
                if next_array_size:
                    fmt_without_num = fmt[-1]  # Get the format type (like 's', 'H', etc.)
                    dynamic_fmt = f"{next_array_size}{fmt_without_num}"

                    logger.debug(f'field_name {field_name}')
                    logger.debug(f'next_array_size {next_array_size}')
                    logger.debug(fmt)
                    logger.debug(data[index:index+field_size])

                    self.values[field_name] = struct.unpack(dynamic_fmt, data[index:index+next_array_size])[0]
                    index+= next_array_size
                    next_array_size = 0
                else:
                    
                    logger.debug(f'field_name {field_name}')
                    logger.debug(fmt)
                    logger.debug(data[index:index+field_size])

                    self.values[field_name] = struct.unpack(fmt, data[index:index+field_size])[0]
                    index+= field_size

    def to_bytes(self) -> bytes:
        packed = bytearray()
        next_array_size = 0
        for field_name, fmt,_ in self.fields:
            val = self.values.get(field_name)
            
            if field_name.startswith("_"):
                # 普通字段，打包并记录大小供下个字段使用
                # print("field_name",field_name,fmt,val)
                packed += struct.pack(fmt, val)
                next_array_size = val
            else:
                if next_array_size:
                    # 使用前面的字段长度生成动态格式
                    fmt_type = fmt[-1]
                    dynamic_fmt = f"{next_array_size}{fmt_type}"
                    if fmt_type == "s":
                        if isinstance(val, str):
                            val = val.encode()
                        val = val.ljust(next_array_size, b'\x00')
                    packed += struct.pack(dynamic_fmt, val)
                    next_array_size = 0
                else:
                    # 非变长字段，正常处理
                    if fmt.endswith("s") and isinstance(val, str):
                        val = val.encode()
                        val = val.ljust(int(fmt[:-1]), b'\x00')
                    packed += struct.pack(fmt, val)
        return bytes(packed)


    def __getattr__(self, name):
        return self.values[name]

    def __setattr__(self, name, value):
        if name in ("fields", "format", "size", "values"):
            super().__setattr__(name, value)
        else:
            self.values[name] = value

    def __str__(self):
        field_strs = []
        for field_name, _, _ in self.fields:
            value = self.values.get(field_name)
            field_strs.append(f"{field_name}={value}")
        return f"DynamicStruct({', '.join(field_strs)})"
    
    def to_json(self):
        json_data = {}
        for field_name, _, _ in self.fields:
            value = self.values.get(field_name)
            # Convert bytes to hex string if value is bytes
            if isinstance(value, bytes):
                value = value.hex()
            json_data[field_name] = value
        return json_data


class BaseValue:
    """基础类型包装类，提供 callback 支持和统一接口"""
    def __init__(self, value, format_char):
        self._value = value
        self._format = format_char
        self.callback = None
    
    @property
    def value(self):
        return self._value
    
    @value.setter
    def value(self, val):
        self._value = val
    
    def parse(self, data: bytes):
        """从字节数据解析值"""
        self._value = struct.unpack(self._format, data[:struct.calcsize(self._format)])[0]
        return struct.calcsize(self._format)
    
    def to_bytes(self) -> bytes:
        """将值转换为字节数据"""
        return struct.pack(self._format, self._value)
    
    def __int__(self):
        return int(self._value)
    
    def __float__(self):
        return float(self._value)
    
    def __str__(self):
        return str(self._value)
    
    def __repr__(self):
        return repr(self._value)
    
    def __eq__(self, other):
        if isinstance(other, BaseValue):
            return self._value == other._value
        return self._value == other
    
    def __ne__(self, other):
        return not self.__eq__(other)
    
    def __lt__(self, other):
        if isinstance(other, BaseValue):
            return self._value < other._value
        return self._value < other
    
    def __le__(self, other):
        if isinstance(other, BaseValue):
            return self._value <= other._value
        return self._value <= other
    
    def __gt__(self, other):
        if isinstance(other, BaseValue):
            return self._value > other._value
        return self._value > other
    
    def __ge__(self, other):
        if isinstance(other, BaseValue):
            return self._value >= other._value
        return self._value >= other
    
    # 支持直接赋值，如 obj.led = 1
    def __setattr__(self, name, value):
        if name in ('_value', '_format', 'callback') and '_value' in self.__dict__:
            if name == '_value':
                object.__setattr__(self, '_value', value)
            elif name == 'callback':
                object.__setattr__(self, 'callback', value)
            else:
                object.__setattr__(self, name, value)
        else:
            object.__setattr__(self, name, value)
    
    # 支持直接获取值（向后兼容）
    def __getattr__(self, name):
        if name == 'value':
            return self._value
        raise AttributeError(f"'{type(self).__name__}' object has no attribute '{name}'")


class CircularBuffer:
    def __init__(self, size):
        self.csize = size
        self.buffer = [None] * size
        self.head = 0
        self.tail = 0
        self.count = 0

    def put(self, data):
        logger.debug(f"Buffer size: {self.csize}, Current count: {self.count}, Data length: {len(data)}")
        if len(data) > self.csize - self.count:
            return -1  # Not enough space in buffer
        for byte in data:
            self.buffer[self.tail] = byte
            self.tail = (self.tail + 1) % self.csize
            self.count += 1
        return 0  # Success

    def get(self, n=1):
        if self.count == 0:
            return -1  # Buffer is empty
        if n > self.count:
            return -1  # Not enough data
        data = []
        for _ in range(n):
            data.append(self.buffer[self.head])
            self.head = (self.head + 1) % self.csize
            self.count -= 1
        return data

    def size(self):
        return self.count

    def clear(self):
        self.head = 0
        self.tail = 0
        self.count = 0
        self.buffer = [None] * self.csize

class FProtocolHeader:
    def __init__(self, from_node=0, to=0, type=0, index=0, data_size=0):
        self.from_node = from_node
        self.to = to
        self.type = type
        self.index = index
        self.data_size = data_size

    @classmethod
    def from_bytes(cls, data):
        if len(data) != 7:
            raise ValueError("Invalid data length for FProtocolHeader")
        from_node = data[0]
        to = data[1]
        type = data[2]
        index = data[3] | (data[4] << 8)
        data_size = data[5] | (data[6] << 8)
        return cls(from_node, to, type, index, data_size)

    def to_bytes(self):
        return [
            self.from_node & 0xFF,
            self.to & 0xFF,
            self.type,
            self.index & 0xFF,
            (self.index >> 8) & 0xFF,
            self.data_size & 0xFF,
            (self.data_size >> 8) & 0xFF
        ]

    def __str__(self):
        return f"FProtocolHeader(from={self.from_node}, to={self.to}, type={self.type}, index={self.index}, data_size={self.data_size})"


class FProtocolType():
    SERVICE_REQUEST_WRITE = 0x01
    SERVICE_REQUEST_READ = 0x02  # Data 0 byte
    SERVICE_RESPONSE_WRITE = 0x03
    SERVICE_RESPONSE_READ = 0x04
    SERVICE_RESPONSE_ERROR = 0x05  # Data 2 bytes -> Error code
    TRANSPORT_DATA = 0x06
    HEART_PING = 0x07
    HEART_PONG = 0x08
    MAX = 0x08
    
class FProtocol:
    FRAME_HEAD = [0x55,0xaa,0x55,0xaa]

    def __init__(self, read_callback, write_callback):
        self.other_nodes = {}  # 其他节点的协议对象
        self.self_node_id = None  # 自己的节点ID
        self.self_node_proto = None  # 自己的协议对象
        self.read_callback = read_callback
        self.write_callback = write_callback
        self.rxbuff = CircularBuffer(2048)
        self.frame = {
            'recv_size': 0,
            'data': [],
            'header': FProtocol.FRAME_HEAD,
            'data_size': 0,
            'fdata': None,
            'error_code': 0
        }
        # 心跳回调函数
        self.heartbeat_callback = None
    
    def add_other_node(self, node_id, node_proto):
        """添加其他节点"""
        self.other_nodes[node_id] = node_proto

    def set_self_node(self, node_id, node_proto):
        """设置自己的节点ID和协议对象"""
        self.self_node_id = node_id
        self.self_node_proto = node_proto
    
    def set_heartbeat_callback(self, callback):
        """设置心跳回调函数"""
        self.heartbeat_callback = callback
    
    def send_heartbeat(self, target_node=None):
        """发送心跳包，默认使用自身节点ID"""
        if target_node is None:
            target_node = self.self_node_id
        return self.fprotocol_write(target_node, FProtocolType.HEART_PING, 0, [])

    def tick(self):
        if self.read_callback:
            data = self.read_callback()
            if data:
                self.rxbuff.put(data)
                logging.debug("put",list(data),self.rxbuff.size())
        
        if self.frame['recv_size'] < 4:
            while self.frame['recv_size'] < 4:
                rdata = self.rxbuff.get(1)
                if rdata == -1:
                    break
                rdata = rdata[0]
                if rdata == FProtocol.FRAME_HEAD[self.frame['recv_size']]:
                    self.frame['data'].append(rdata)
                    self.frame['recv_size'] += 1
                elif rdata == FProtocol.FRAME_HEAD[0]:
                    self.frame['data'] = [rdata]
                    self.frame['recv_size'] = 1
                else:
                    self.frame['recv_size'] = 0
        
        if self.frame['recv_size'] >= 4 and self.frame['recv_size'] < 11:
            additional_data = self.rxbuff.get(11 - self.frame['recv_size'])
            if additional_data == -1:
                return
            self.frame['data'].extend(additional_data)
            self.frame['recv_size'] = len(self.frame['data'])
            if self.frame['recv_size'] != 11:
                return
            header = self.parse_header(self.frame['data'][4:11])
            self.frame['header']  = header
            logger.debug(f"header={header}")
            if header.to == self.self_node_id:
                logger.debug(f"Recv data for self node {header.to}")
            elif header.from_node in self.other_nodes.keys():
                if header.type == FProtocolType.HEART_PING:
                    self.frame['data_size'] = 0
                    self.frame['fdata'] = None
                    logger.debug(f"Recv data from node {self.frame['header'].from_node} data_size={self.frame['header'].data_size}")
                else:
                    # fdata = self.slave_nodes[self.frame['header'].node].get_index_data(self.frame['header'].index)
                    # self.frame['fdata'] = fdata
                    if header.type == FProtocolType.SERVICE_RESPONSE_ERROR:
                        self.frame['data_size'] = self.frame['header'].data_size  # 2
                    elif header.type == FProtocolType.SERVICE_REQUEST_WRITE:
                        self.frame['data_size'] = self.frame['header'].data_size # 0
                    elif header.type == FProtocolType.SERVICE_RESPONSE_READ:
                        self.frame['data_size'] = self.frame['header'].data_size 
                    elif header.type == FProtocolType.TRANSPORT_DATA:
                        self.frame['data_size'] = self.frame['header'].data_size 
                    else:
                        self.frame['recv_size'] = 0 # 重新接收
            # print(self.frame['recv_size'],self.frame['data_size'])
            else:
                self.frame['recv_size'] = 0
                
        if self.frame['recv_size'] >= 11 and self.frame['recv_size'] < (11 + self.frame['data_size']):
            additional_data = self.rxbuff.get(self.frame['data_size'])
            if additional_data == -1:
                return
            self.frame['data'].extend(additional_data)
            self.frame['recv_size'] = len(self.frame['data'])
            logger.debug(f"Recv data from node {self.frame['header'].node} data_size={self.frame['data_size']}")
            # print()
        
        if self.frame['recv_size'] >= (11 + self.frame['data_size']):
            checksum_data = self.rxbuff.get(2) # checksum
            if checksum_data == -1:
                return
            # print("self.frame['recv_size']",self.frame['recv_size'],self.frame['data'],checksum_data)
            calculated_checksum = self.checksum16(self.frame['data'])
            received_checksum = checksum_data[0] << 8 | checksum_data[1]
            if calculated_checksum == received_checksum:
                if self.frame['header'].type == FProtocolType.SERVICE_RESPONSE_ERROR:
                    self.frame['error_code'] = self.frame['data'][11] << 8 | self.frame['data'][12]
                self.process_frame(self.frame)  

            self.frame['data'].clear()
            self.frame['recv_size'] = 0
            # print("self.frame['recv_size']",self.frame['recv_size'])
        if self.frame['recv_size'] > 1024:
            self.frame['recv_size'] = 0


    def process_frame(self, frame):
        header = frame['header']
        fdata = None
        
        # 处理心跳
        if header.type == FProtocolType.HEART_PING and header.from_node in self.other_nodes.keys():
            # 发送心跳响应
            if self.heartbeat_callback:
                self.heartbeat_callback(header.type, header.from_node, 0)
            self.fprotocol_write(header.from_node, FProtocolType.HEART_PONG, 0, [])
            print(f"[Heartbeat] 收到来自节点 {header.from_node} 的心跳，已回复")

        # 处理其他类型的消息
        elif header.from_node in self.other_nodes.keys():
            if header.type == FProtocolType.SERVICE_RESPONSE_READ or header.type == FProtocolType.TRANSPORT_DATA:
                fdata = self.other_nodes[self.frame['header'].from_node].get_index_data(self.frame['header'].index)
                fdata.parse(bytes(frame['data'][11:]))

            if fdata and fdata.callback:
                fdata.callback(header.type, fdata, frame['error_code'])


    def fprotocol_write(self, to, type, index, data, data_size=0):
        frame = []
        frame.extend(FProtocol.FRAME_HEAD)
        frame.extend([
            self.self_node_id & 0xFF,
            to & 0xFF,
            type,
            index & 0xFF,
            (index >> 8) & 0xFF,
            data_size & 0xFF,
            (data_size >> 8) & 0xFF
        ])
        if data:
            frame.extend(data)
        checksum = self.checksum16(frame)
        frame.append(checksum >> 8 & 0xFF)
        frame.append(checksum & 0xFF)
        self.write_callback(bytes(frame))

    def checksum16(self,data):
        sum_val = 0
        for byte in data:
            sum_val += byte
            # Handle carry (keep sum within 16 bits)
            if sum_val & 0xFFFF0000:
                sum_val = (sum_val & 0xFFFF) + (sum_val >> 16)
        # Return one's complement of the lower 16 bits
        return (~sum_val & 0xFFFF)

    def parse_header(self,data):
        head = FProtocolHeader.from_bytes(data)
        # print("parse_header",head,head.to_bytes())
        return head
    
    def read_put(self, data):
        result = self.rxbuff.put(data)
        if result == -1:
            logger.debug("Buffer is full, cannot write data")
        else:
            logger.debug(f"Data written to buffer: {data}")
            self.tick()

File: src/fprotocol/fprotocol.cpp
#include "fprotocol.hpp"
#include <algorithm>
#include <iostream>

// #define DEBUG 1

namespace FProtocol {

const uint8_t Handler::FRAME_HEAD[4] = {0x55, 0xAA, 0x55, 0xAA};

// ProtocolData Implementation
ProtocolData::ProtocolData(uint16_t idx, uint16_t size, void* data_ptr, 
                          CallbackFunction cb, 
                          std::shared_ptr<StructDescriptor> desc)
    : index_(idx), data_size_(size), data_(data_ptr), callback_(cb), struct_desc_(desc) {
}

// RingBuffer Implementation
RingBuffer::RingBuffer(uint32_t size) 
    : buf_(std::make_unique<uint8_t[]>(size)), buf_size_(size), p_in_(0), p_out_(0) {
}

RingBuffer::~RingBuffer() = default;

bool RingBuffer::put(const uint8_t* buf, uint32_t size) {
    if (size + this->size() >= buf_size_) {
        return false;
    }
    
    for (uint32_t i = 0; i < size; i++) {
        buf_[p_in_] = buf[i];
        p_in_ = (p_in_ + 1) % buf_size_;
    }
    return true;
}

uint32_t RingBuffer::get(uint8_t* buf, uint32_t size) {
    if (p_in_ == p_out_) {
        return 0;
    }
    if (this->size() < size) {
        return 0;
    }
    
    uint32_t i;
    for (i = 0; i < size; i++) {
        buf[i] = buf_[p_out_];
        p_out_ = (p_out_ + 1) % buf_size_;
        
        if (p_in_ == p_out_) {
            return i + 1;
        }
    }
    return i;
}

uint32_t RingBuffer::size() const {
    if (p_in_ >= p_out_) {
        return p_in_ - p_out_;
    } else {
        return buf_size_ - (p_out_ - p_in_);
    }
}

void RingBuffer::clear() {
    p_in_ = 0;
    p_out_ = 0;
}

// Frame Implementation
Frame::Frame() 
    : data_size_(0), recv_size_(0), from_(0), error_code_(0), 
      protocol_data_(nullptr), data_(std::make_unique<uint8_t[]>(FPROTOCOL_FRAME_DATA_SIZE)) {
    std::memset(&header_, 0, sizeof(header_));
}

void Frame::reset() {
    recv_size_ = 0;
    data_size_ = 0;
    from_ = 0;
    error_code_ = 0;
    protocol_data_ = nullptr;
    std::memset(&header_, 0, sizeof(header_));
}

// Handler Implementation
Handler::Handler(ReadFunction read_func, WriteFunction write_func)
    : self_node_id_(0), read_func_(read_func), write_func_(write_func),
      frame_(std::make_unique<Frame>()), 
      rx_buffer_(std::make_unique<RingBuffer>(RING_BUFFER_SIZE)),
      other_node_count_(0) {
    other_node_table_.reserve(MAX_NODE);
    other_node_info_table_.reserve(MAX_NODE);
}

void Handler::setSelfNode(uint8_t node, GetIndexInfoFunction get_index_info) {
    self_node_id_ = node;
    get_node_info_ = get_index_info;
}

void Handler::addOtherNode(uint8_t node, GetIndexInfoFunction get_index_info) {
    if (other_node_count_ >= MAX_NODE) {
        return;
    }
    other_node_table_.push_back(node);
    other_node_info_table_.push_back(get_index_info);
    other_node_count_++;
}

ProtocolData* Handler::getOtherNodeData(uint8_t node, uint16_t index) {
    if (other_node_count_ == 0) {
        return nullptr;
    }
    
    for (size_t i = 0; i < other_node_count_; i++) {
        if (other_node_table_[i] == node) {
            return other_node_info_table_[i](index);
        }
    }
    return nullptr;
}

void Handler::readPut(const uint8_t* buf, uint32_t size) {
    rx_buffer_->put(buf, size);
#ifdef DEBUG
    std::printf("put data to ring_buffer_size: %d\n", size);
#endif
}

void Handler::tick() {
    static uint8_t data[256];
    static uint8_t from = 0;
    uint8_t rdata;
    
    if (read_func_) {
        int16_t rsize = read_func_(from, data, 256);
        if (rsize > 0) {
            rx_buffer_->put(data, rsize);
#ifdef DEBUG
            std::printf("put data to ring_buffer_size: %d\n", rsize);
#endif
        }
    }

#ifdef DEBUG
    std::printf("ring_size: %d frame->recv_size:%u\n", rx_buffer_->size(), frame_->getRecvSize());
#endif

    if (frame_->getRecvSize() < 4) {
        while (frame_->getRecvSize() < 4) {
            if (rx_buffer_->get(&rdata, 1)) {
#ifdef DEBUG
                std::printf("rdata: %02x\n", rdata);
#endif
                if (rdata == FRAME_HEAD[frame_->getRecvSize()]) {
                    frame_->getData()[frame_->getRecvSize()] = rdata;
                    frame_->addRecvSize(1);
                } else if (rdata == FRAME_HEAD[0]) {
                    frame_->getData()[0] = rdata;
                    frame_->setRecvSize(1);
                } else {
                    frame_->setRecvSize(0);
                }
            } else {
                break;
            }
        }
    }

    const uint16_t frame_header_size = sizeof(FRAME_HEAD) + sizeof(ProtocolHeader);

    if (frame_->getRecvSize() >= 4 && frame_->getRecvSize() < frame_header_size) {
        uint32_t read_size = rx_buffer_->get(frame_->getData() + sizeof(FRAME_HEAD), sizeof(ProtocolHeader));
        frame_->addRecvSize(read_size);

        if (frame_->getRecvSize() != frame_header_size) {
            return;
        } else {
            ProtocolHeader header;
            std::memcpy(&header, frame_->getData() + sizeof(FRAME_HEAD), sizeof(ProtocolHeader));
            frame_->setHeader(header);
            frame_->setFrom(header.from);

#ifdef DEBUG
            std::printf("Header Info - From: %02x, To: %02x, Type: %02x, Index: %d, DSize: %d \n",
                       header.from, header.to, header.type, header.index, header.data_size);
            std::printf("%d self_node_id: %02x\n", __LINE__, self_node_id_);
#endif

            // Check if node ID is self
            if (header.to == self_node_id_) {
                switch (static_cast<ProtocolType>(header.type)) {
                case ProtocolType::HEART_PONG:
                    frame_->setDataSize(0);
                    break;
                case ProtocolType::SERVICE_RESPONSE_ERROR:
                    frame_->setDataSize(header.data_size);
                    break;
                case ProtocolType::SERVICE_REQUEST_READ:
                case ProtocolType::SERVICE_RESPONSE_READ:
                case ProtocolType::TRANSPORT_DATA:
                    frame_->setDataSize(header.data_size);
                    if (get_node_info_) {
                        frame_->setProtocolData(get_node_info_(header.index));
                        if (frame_->getProtocolData() == nullptr) {
#ifdef DEBUG
                            std::printf("No this index\n");
#endif
                            frame_->setDataSize(0);
                            frame_->setRecvSize(0);
                        }
                    }
                    break;
                default:
                    frame_->setDataSize(0);
                    frame_->setRecvSize(0);
                    break;
                }
            }
            // Check if message from other nodes
            else if (getOtherNodeData(header.from, header.index) != nullptr) {
                frame_->setProtocolData(getOtherNodeData(header.from, header.index));
                switch (static_cast<ProtocolType>(header.type)) {
                case ProtocolType::SERVICE_RESPONSE_ERROR:
                    frame_->setDataSize(header.data_size);
                    break;
                case ProtocolType::SERVICE_REQUEST_WRITE:
                    frame_->setDataSize(header.data_size);
                    break;
                case ProtocolType::SERVICE_RESPONSE_READ:
                case ProtocolType::TRANSPORT_DATA:
                    frame_->setDataSize(header.data_size);
                    break;
                case ProtocolType::HEART_PONG:
                    frame_->setDataSize(0);
                    break;
                default:
                    frame_->setDataSize(0);
                    frame_->setRecvSize(0);
                    break;
                }
            } else {
                frame_->setDataSize(0);
                frame_->setRecvSize(0);
            }
        }
    }

    if (frame_->getRecvSize() >= frame_header_size && frame_->getRecvSize() < (frame_header_size + frame_->getDataSize())) {
        uint32_t read_size = rx_buffer_->get(frame_->getData() + frame_->getRecvSize(), frame_->getDataSize());
        frame_->addRecvSize(read_size);
    }

    if (frame_->getRecvSize() >= frame_header_size + frame_->getDataSize()) {
        uint32_t read_size = rx_buffer_->get(frame_->getData() + frame_->getRecvSize(), 2);
        frame_->addRecvSize(read_size);

        if (frame_->getRecvSize() == (frame_header_size + frame_->getDataSize() + 2)) {
            uint16_t calculated_checksum = checksum16(frame_->getData(), frame_header_size + frame_->getDataSize());
            uint16_t received_checksum = (frame_->getData()[frame_header_size + frame_->getDataSize()] << 8) |
                                       frame_->getData()[frame_header_size + frame_->getDataSize() + 1];
            if (calculated_checksum == received_checksum) {
                processRequest();
            }
            frame_->setRecvSize(0);
        }
    }

    if (frame_->getRecvSize() > FPROTOCOL_FRAME_DATA_SIZE) {
        frame_->setRecvSize(0);
    }
}

void Handler::processRequest() {
    static uint8_t send_buff[FPROTOCOL_FRAME_DATA_SIZE];
    uint16_t ret = 0;
    std::memcpy(send_buff, FRAME_HEAD, 4);
    
    const auto& header = frame_->getHeader();
    const uint16_t frame_header_size = sizeof(FRAME_HEAD) + sizeof(ProtocolHeader);

    // Slave node
    if (header.to == self_node_id_) {
#ifdef DEBUG
        std::printf("Type: %02x\n", header.type);
        std::printf("frame->header.from: %02x frame.protocol_data: %p\n", header.from, frame_->getProtocolData());
#endif
        if (static_cast<ProtocolType>(header.type) == ProtocolType::HEART_PONG) {
            if (heart_ping_callback_) {
                heart_ping_callback_(header.from);
            }
            return;
        }
        
        // Need to write data
        if (static_cast<ProtocolType>(header.type) == ProtocolType::SERVICE_REQUEST_WRITE ||
            static_cast<ProtocolType>(header.type) == ProtocolType::TRANSPORT_DATA) {
            if (frame_->getProtocolData() && frame_->getProtocolData()->getStructDescriptor()) {
                unpackStruct(frame_->getData() + frame_header_size, frame_->getProtocolData()->getData(),
                           *frame_->getProtocolData()->getStructDescriptor());
            }
        }
        
        // Callback
        if (frame_->getProtocolData() && frame_->getProtocolData()->getCallback()) {
            ret = frame_->getProtocolData()->getCallback()(header.type, frame_->getFrom(), frame_->getErrorCode());
        }
        
        // Need to reply SERVICE_REQUEST_WRITE
        if (static_cast<ProtocolType>(header.type) == ProtocolType::SERVICE_REQUEST_WRITE) {
            if (ret == 0) {
                write(header.from, ProtocolType::SERVICE_RESPONSE_WRITE, header.index, nullptr, 0);
            } else {
                write(header.from, ProtocolType::SERVICE_RESPONSE_ERROR, header.index, &ret, 2);
            }
        }

        // Need to reply SERVICE_REQUEST_READ
        if (static_cast<ProtocolType>(header.type) == ProtocolType::SERVICE_REQUEST_READ) {
            if (frame_->getProtocolData()) {
                write(header.from, ProtocolType::SERVICE_RESPONSE_READ, header.index,
                     frame_->getProtocolData()->getData(), frame_->getProtocolData()->getDataSize(),
                     frame_->getProtocolData()->getStructDescriptor());
            }
        }
    }
    else { // Master node
        // Need to write data
        if (static_cast<ProtocolType>(header.type) == ProtocolType::SERVICE_RESPONSE_READ ||
            static_cast<ProtocolType>(header.type) == ProtocolType::TRANSPORT_DATA) {
            std::memcpy(frame_->getProtocolData()->getData(), frame_->getData() + frame_header_size, frame_->getDataSize());
        }
        
        // Callback
        if (frame_->getProtocolData() && frame_->getProtocolData()->getCallback()) {
            ret = frame_->getProtocolData()->getCallback()(header.type, frame_->getFrom(), frame_->getErrorCode());
        }
    }
}

size_t Handler::packStruct(uint8_t* buffer, const void* data, const StructDescriptor& desc) {
    size_t offset = 0;
    for (const auto& field : desc.fields) {
        const uint8_t* field_ptr = static_cast<const uint8_t*>(data) + field.offset;
        size_t len = 0;

        if (field.size_field >= 0) {
            // Variable length array
            const auto& size_field = desc.fields[field.size_field];
            const uint8_t* size_ptr = static_cast<const uint8_t*>(data) + size_field.offset;
            len = *size_ptr;
        } else if (field.array_len > 0) {
            // Fixed length array
            len = field.array_len;
        } else {
            // Single value field
            switch (field.type) {
            case FieldType::UINT8:
                len = 1;
                break;
            case FieldType::UINT16:
                len = 2;
                break;
            case FieldType::UINT32:
                len = 4;
                break;
            case FieldType::FLOAT:
                len = 4;
                break;
            default:
                break;
            }
        }

        std::memcpy(buffer + offset, field_ptr, len);
        offset += len;
    }
    return offset;
}

size_t Handler::unpackStruct(const uint8_t* buffer, void* data, const StructDescriptor& desc) {
#ifdef DEBUG
    std::printf("unpack_struct ");
#endif
    size_t offset = 0;
    for (const auto& field : desc.fields) {
        uint8_t* field_ptr = static_cast<uint8_t*>(data) + field.offset;
        size_t len = 0;

        if (field.size_field >= 0) {
            // Variable length array
            const auto& size_field = desc.fields[field.size_field];
            const uint8_t* size_ptr = static_cast<const uint8_t*>(data) + size_field.offset;
            len = *size_ptr;
        } else if (field.array_len > 0) {
            len = field.array_len;
        } else {
            switch (field.type) {
            case FieldType::UINT8:
                len = 1;
                break;
            case FieldType::UINT16:
                len = 2;
                break;
            case FieldType::UINT32:
                len = 4;
                break;
            case FieldType::FLOAT:
                len = 4;
                break;
            default:
                break;
            }
        }

        std::memcpy(field_ptr, buffer + offset, len);
        offset += len;
    }
    return offset;
}

uint16_t Handler::write(uint8_t to, ProtocolType type, uint16_t index,
                       const void* data, uint16_t size,
                       std::shared_ptr<StructDescriptor> desc) {
#ifdef DEBUG
    std::printf("write size:%d\n", size);
#endif
    static uint8_t send_buff[FPROTOCOL_FRAME_DATA_SIZE];
    std::memcpy(send_buff, FRAME_HEAD, 4);

    ProtocolHeader header;
    const uint16_t frame_header_size = sizeof(FRAME_HEAD) + sizeof(ProtocolHeader);
    header.from = self_node_id_;
    header.to = to;
    header.type = static_cast<uint8_t>(type);
    header.index = index;

    if (size && data) {
        if (desc) {
            size = packStruct(send_buff + frame_header_size, data, *desc);
        } else {
            std::memcpy(send_buff + frame_header_size, data, size);
        }
    }

    header.data_size = size;
    std::memcpy(send_buff + 4, &header, sizeof(ProtocolHeader));

#ifdef DEBUG
    std::printf("data_size: %02x\n", header.data_size);
#endif

    uint16_t checksum = checksum16(send_buff, frame_header_size + size);
    send_buff[frame_header_size + size] = checksum >> 8;
    send_buff[frame_header_size + size + 1] = checksum & 0xFF;

    return write_func_(to, send_buff, frame_header_size + size + 2);
}

int8_t Handler::heartPing() {
    return write(self_node_id_, ProtocolType::HEART_PING, 0, nullptr, 0) > 0;
}

void Handler::setHeartPingCallback(HeartPingCallback callback) {
    heart_ping_callback_ = callback;
}

uint16_t Handler::checksum16(const uint8_t* data, size_t length) {
    uint32_t sum = 0;
    for (size_t i = 0; i < length; i++) {
        sum += data[i];
        if (sum & 0xFFFF0000) {
            sum = (sum & 0xFFFF) + (sum >> 16);
        }
    }
    return ~(sum & 0xFFFF);
}

} // namespace FProtocol